src/language-js/embed.js

```diff
@@ -255,62 +365,112 @@ function transformCssDoc(quasisDoc, path, print) {
+  string.slice(-CSS_EXTRA_SEMICOLON_MARK_LENGTH) === CSS_EXTRA_SEMICOLON_MARK;
+const hasCSSPlaceholder = doc => {
+  if (!doc || !doc.parts) {
+    return false;
+  }
+  const text = doc.parts.filter(part => typeof part === "string").join();
+  return (
+    [CSS_PROPERTY_PLACEHOLDER, CSS_EXTRA_SEMICOLON_MARK].some(
+      placeholder => text.indexOf(placeholder) !== -1
+    ) || cssPlaceholder.hasPlaceholder(text)
+  );
+};
 function replacePlaceholders(quasisDoc, expressionDocs) {
   if (!expressionDocs || !expressionDocs.length) {
     return quasisDoc;
   }
 
-  const expressions = expressionDocs.slice();
-  let replaceCounter = 0;
-  const newDoc = mapDoc(quasisDoc, doc => {
-    if (!doc || !doc.parts || !doc.parts.length) {
+  const replacedExpressions = [];
+  const restoredDoc = mapDoc(quasisDoc, doc => {
+    if (!hasCSSPlaceholder(doc)) {
       return doc;
     }
-    let parts = doc.parts;
-    const atIndex = parts.indexOf("@");
-    const placeholderIndex = atIndex + 1;
-    if (
-      atIndex > -1 &&
-      typeof parts[placeholderIndex] === "string" &&
-      parts[placeholderIndex].startsWith("prettier-placeholder")
-    ) {
-      // If placeholder is split, join it
-      const at = parts[atIndex];
-      const placeholder = parts[placeholderIndex];
-      const rest = parts.slice(placeholderIndex + 1);
-      parts = parts
-        .slice(0, atIndex)
-        .concat([at + placeholder])
-        .concat(rest);
-    }
-    const atPlaceholderIndex = parts.findIndex(
-      part =>
-        typeof part === "string" && part.startsWith("@prettier-placeholder")
-    );
-    if (atPlaceholderIndex > -1) {
-      const placeholder = parts[atPlaceholderIndex];
-      const rest = parts.slice(atPlaceholderIndex + 1);
-      const placeholderMatch = placeholder.match(
-        /@prettier-placeholder-(.+)-id([\s\S]*)/
-      );
-      const placeholderID = placeholderMatch[1];
-      // When the expression has a suffix appended, like:
-      // animation: linear ${time}s ease-out;
-      const suffix = placeholderMatch[2];
-      const expression = expressions[placeholderID];
-
-      replaceCounter++;
-      parts = parts
-        .slice(0, atPlaceholderIndex)
-        .concat(["${", expression, "}" + suffix])
-        .concat(rest);
-    }
+
+    const parts = doc.parts.slice().reduce((parts, part, index, original) => {
+      if (typeof part !== "string") {
+        parts.push(part);
+        return parts;
+      }
+
+      // clean extra semicolon mark
+      if (endsWithCSSSemicolonMark(part)) {
+        part = part.slice(0, -CSS_EXTRA_SEMICOLON_MARK_LENGTH);
+        original[index] = part;
+
+        // find following semicolon
+        let semicolonIndex = -1;
+        for (let i = index + 1; i < original.length; i++) {
+          const value = original[i];
+          if (typeof value !== "string" || !value.trim()) {
+            continue;
+          }
+          if (value !== ";") {
+            throw new Error(
```

Comment:
Very bad idea, we should not throw the error on broken code, just do not format it

Reply:
this whole function is `try/catch` ed

Reply:
hm, can't find try/catch here

Reply:
https://github.com/prettier/prettier/blob/64ab703d4122c2c835741dcf888b3df663c33c2a/src/main/ast-to-doc.js#L109

Reply:
also, this error is put there for safety, actually I never find a way to trigger it, unless you modify ast manually

Reply:
:+1: